agent/src/main/java/com/teamscale/jacoco/agent/testimpact/TestwiseCoverageAgent.java

```diff
@@ -5,46 +5,117 @@
+ * 
+ * If the port given in the agent options is occupied, it will use the next
+ * available free port for itself and try to register with a primary agent that
+ * listens at the original port. Secondary agents will unregister themselves
+ * with the primary agent before shutting down.
  */
 public class TestwiseCoverageAgent extends AgentBase {
 
-	/** Path parameter placeholder used in the http requests. */
+	/** The maximum port number for the agent to listen at. */
+	private static final int MAX_PORT_NUMBER = 65535;
+
+	/** Path parameter placeholder used in the HTTP requests. */
 	private static final String TEST_ID_PARAMETER = ":testId";
 
+	/** Map of port number to secondary agent service. */
+	private final Map<Integer, IAgentService> secondaryAgents = new LinkedHashMap<>();
+
+	/** Primary agent service (if any) or null. */
+	private IAgentService primaryAgent = null;
+
 	/** The agent options. */
 	private AgentOptions options;
 
+	/** The service for the HTTP API. */
+	private final Service service;
+
 	/** Constructor. */
-	public TestwiseCoverageAgent(AgentOptions options) throws IllegalStateException {
-		super(options);
+	public TestwiseCoverageAgent(AgentOptions options, IAgent agent) throws IOException {
+		super(options, agent);
 		this.options = options;
-		initServer();
+		service = Service.ignite();
+		initServer(determineFreePort());
 	}
 
 	/**
-	 * Starts the http server, which waits for information about started and finished tests.
+	 * Starts the HTTP server, which waits for information about started and
+	 * finished tests.
+	 * 
+	 * @throws IOException
+	 *             in case registration with the primary agent failed.
 	 */
-	private void initServer() {
-		logger.info("Listening for test events on port {}.", options.getHttpServerPort());
-		port(options.getHttpServerPort());
+	private void initServer(int port) throws IOException {
+		logger.info("Listening for test events on port {}.", port);
+
+		service.initExceptionHandler(this::logInitException);
+		service.port(port);
+
+		service.get("/test", (request, response) -> controller.getSessionId());
+
+		service.post("/test/start/" + TEST_ID_PARAMETER, this::handleTestStart);
+		service.post("/test/end/" + TEST_ID_PARAMETER, this::handleTestEnd);
+
+		service.post("/register", this::handleRegister);
+		service.delete("/register", this::handleUnregister);
+
+		if (port != options.getHttpServerPort()) {
+			registerWithPrimaryAgent();
+		}
+
+		service.awaitInitialization();
+	}
 
-		get("/test", (request, response) -> controller.getSessionId());
+	/**
+	 * Returns the lowest unused local port that is equal or greater than the
+	 * one provided in {@link #options}.
+	 */
+	private int determineFreePort() throws IOException {
+		for (int currentPort = options.getHttpServerPort(); currentPort < MAX_PORT_NUMBER;) {
+			try (ServerSocket s = new ServerSocket(currentPort)) {
+				// found a free port. Jump out
+				return currentPort;
```

Comment:
Do we already reserve the port by creating a ServerSocket here?
We will run into race conditions with other agents.
If the port is already reserved for us, then please document here.

Reply:
This is a known issue. We do *not* reserve the port, because there is no way to do instantiate the spark service with an already established server socket. So we need to release it and hope that it is still available once the integrated Jetty tries to bind to the same port. Asked the customer to check whether this is an issue in practice. Will add a comment to the code to make that clear.

Reply:
Also added a hard fail, so we do not lose data.